### PR TITLE
py3 support

### PR DESCRIPTION
**Pull request by rredburn**
_What:_

Add Python 3 compatibility.

_Why:_

Python 3, now in it's fifth revision, has obtained enough critical mass in the developer community to make supporting it worthwhile.  Bluemix supports Python 3 Cloud Foundry applications as a built in community runtime and makes attaching a Cloudant service an easy and desirable option.  Lack of a Python 3 compatible API inhibits adoption of the Cloudant service. 

_How:_

A new private module, _2to3.py, was added containing compatibility methods.  py2 or py3 unique types and methods were replaced throughout the code base with equivalents from the compatibility module.  Python 2 is treated as the special case given it's unique semantics are "deprecated" and py3 and onward is the new normal (all hail the BDFL).  

_Tests:_

All test cases under the tests/ directory were updated to function under py2 and py3 with similar compatibility methods.  All unit tests (tests/unit/*) were run and pass with py2.7 and py3.5 using a local couchdb instance.  The 85 cloudant specific test cases have not yet been run to my knowledge.  The Travis CI environment only executes a py2.7 test -- a py3.5 test should be added before a final merge is complete (I'm not sure how to do it).

_Issues:_

#23 

reviewer @alfinkel
reviewer @evansde77

### CODE REVIEW

**Comment by rredburn** [comment]
ok, so pylint with py2.7 fails on py3.5 conditional imports.  I was able to get pylint/3.5 to ignore the py2 imports with local overrides, but pylint/2.7 won't take them for some reason.  
 

> $ pylint --rcfile=../pylintrc cloudant
> ************* Module cloudant.credentials
> E: 28, 4: Unable to import 'configparser' (import-error)
> ************* Module cloudant.database
> E: 31, 4: No name 'parse' in module 'urllib' (no-name-in-module)
> E: 31, 4: Unable to import 'urllib.parse' (import-error)
> ************* Module cloudant.document
> E: 30, 4: No name 'parse' in module 'urllib' (no-name-in-module)
> E: 30, 4: Unable to import 'urllib.parse' (import-error)

For example, configparser in credentials is overridden but seemingly ignored by pylint/2.7.  Anyone know the secret sauce?

```python
# pylint: disable=wrong-import-order
from ._2to3 import PY2
if PY2:
    # pylint: disable=wrong-import-order
    from ConfigParser import RawConfigParser
else:
    # pylint: disable=wrong-import-order, import-error
    from configparser import RawConfigParser
`

**Comment by rredburn** [comment]
found a fix with line local vs. block pylint overrides.  First time I've seen Travis integration on a Python library, pretty slick.   You'll have to add a py3 check to it.

**Comment by alfinkel** [comment]
@rredburn - At first glance this looks great.  I just sent you an internal email outlining what needs to happen in order for you to be able to contribute to this repo.  It is a fairly simple process.

**Comment by alfinkel** [comment]
@rredburn:  Can you edit your original comment to follow our PR submission process initial comment.  Take a look at https://github.com/cloudant/python-cloudant/pull/89 as a reference.  In your edit reference issue #23, which is the Python3 support issue number, in your comment and add me as a reviewer.  Once I have a look and provide comments, I'll get one other member from our team to take a look as well.

File: src/cloudant/document.py [hunk 1]
```diff
@@ -17,13 +17,19 @@
 """
 import json
 import posixpath
-import urllib
 import requests
-
 from requests.exceptions import HTTPError
 
+# pylint: disable=wrong-import-order
```
Comment (evansde77) [inline src/cloudant/document.py]: given this stanza appears a couple of times and I think string quoting is going to be used in other places, may be cleaner to set up a quote_plus_ style symbol in the _2to3 module.

**Comment by evansde77** [comment]
This looks good to me, would suggest dropping a python 3 build in the .travis.yml on this branch and letting travis give it a wallop as well. 
(See https://github.com/cloudant/python-cloudant/blob/master/.travis.yml#L4) 
Eg: 
https://github.com/evansde77/pluggage/blob/develop/.travis.yml#L5

The suggestion about defining a quote_plus_ in _2to3 is at your discretion, but I would prefer to keep the compat stuff in one place and avoid copypasta in other modules. 

+1

**Comment by alfinkel** [comment]
Add the python version `- "3.5"` after the `- "2.7"`entry in the travis.yml (https://github.com/cloudant/python-cloudant/blob/master/.travis.yml#L4) as @evansde77 suggests.  As in:

```
python:
  - "2.7"
  - "3.5"
```

This way we can see how Travis likes it.

**Comment by alfinkel** [comment]
It would also be great if you could confirm that the Cloudant specific tests also work in Python 3.  You can follow instructions [here](https://github.com/cloudant/python-cloudant/blob/master/tests/unit/db/unit_t_db_base.py#L22).  Basically it is a matter of setting 4 environment variables and kicking off the tests as you would normally.

File: src/cloudant/_2to3.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2015 IBM. All rights reserved.
```
Comment (alfinkel) [inline src/cloudant/_2to3.py]: Change to 2016

File: src/cloudant/changes.py [hunk 3]
```diff
@@ -19,6 +19,9 @@
 
 import json
 
+from ._2to3 import next_, unicode_
+
+
```
Comment (alfinkel) [inline src/cloudant/changes.py]: Remove one blank line.

File: src/cloudant/database.py [hunk 4]
```diff
@@ -18,9 +18,17 @@
...
+
 from requests.exceptions import HTTPError
 
+# pylint: disable=wrong-import-order
+from ._2to3 import PY2
+if PY2:
+    # pylint: disable=wrong-import-order,no-name-in-module
+    from urllib import quote_plus
+else:
+    from urllib.parse import quote_plus  # pylint: disable=import-error,no-name-in-module
```
Comment (alfinkel) [inline src/cloudant/database.py]: Can you put the pylint disable comment above the statement?  This more closely matches the coding standard that we've adopted.

File: src/cloudant/database.py [hunk 5]
```diff
@@ -586,11 +590,8 @@ def missing_revisions(self, doc_id, *revisions):
         url = posixpath.join(self.database_url, '_missing_revs')
         data = {doc_id: list(revisions)}
 
-        resp = self.r_session.post(
-            url,
-            headers={'Content-Type': 'application/json'},
-            data=json.dumps(data)
-        )
+        resp = self.r_session.post(url, headers={'Content-Type': 'application/json'},
+                                   data=json.dumps(data))
```
Comment (alfinkel) [inline src/cloudant/database.py]: Can you return the statement to its original form?  Again, the coding standard we've adopted.

File: src/cloudant/database.py [hunk 6]
```diff
@@ -615,11 +616,8 @@ def revisions_diff(self, doc_id, *revisions):
         url = posixpath.join(self.database_url, '_revs_diff')
         data = {doc_id: list(revisions)}
 
-        resp = self.r_session.post(
-            url,
-            headers={'Content-Type': 'application/json'},
-            data=json.dumps(data)
-        )
+        resp = self.r_session.post(url, headers={'Content-Type': 'application/json'},
+                                   data=json.dumps(data))
```
Comment (alfinkel) [inline src/cloudant/database.py]: Return statement to original form.

File: src/cloudant/document.py [hunk 7]
```diff
@@ -17,13 +17,19 @@
...
+from ._2to3 import PY2, unicode_
+if PY2:
+    from urllib import quote, quote_plus  # pylint: disable=no-name-in-module
+else:
+    from urllib.parse import quote, quote_plus  # pylint: disable=import-error,no-name-in-module
```
Comment (alfinkel) [inline src/cloudant/document.py]: Both pylint disable comments should be above their referenced statements.

File: src/cloudant/replicator.py [hunk 8]
```diff
@@ -163,14 +163,11 @@ def update_state():
...
-                state = repl_doc.get('_replication_state')
+                arepl_doc = self.database[repl_id]
+                arepl_doc.fetch()
+                return arepl_doc, arepl_doc.get('_replication_state')
             except KeyError:
-                repl_doc = None
-                state = None
-
-            return repl_doc, state
+                return None, None
```
Comment (alfinkel) [inline src/cloudant/replicator.py]: Any reason why you chose to rewrite this try block?  It doesn't appear to have any py2 v py3 issues.  Also why the renaming of `repl_doc` to `arepl_doc`?

File: src/cloudant/result.py [hunk 9]
```diff
@@ -360,10 +360,11 @@ def __getitem__(self, key):
 
         :returns: Rows data in JSON format
         """
-        if 'skip' in self.options or 'skip' in self._ref.keys():
+        refkeys = set(self._ref.keys())
+        if 'skip' in self.options or 'skip' in refkeys:
             msg = 'Cannot use skip parameter with QueryResult slicing.'
             raise CloudantArgumentError(msg)
-        if 'limit' in self.options or 'limit' in self._ref.keys():
+        if 'limit' in self.options or 'limit' in refkeys:
```
Comment (alfinkel) [inline src/cloudant/result.py]: Why the introduction of `refkeys` in this series of if statements?  This does not seem related to adding py3 support.

File: tests/unit/db/account_tests.py [hunk 10]
```diff
@@ -232,14 +234,14 @@ def test_delete_remote_db_via_delitem(self):
...
-            db = self.client.create_database(dbname)
+            _ = self.client.create_database(dbname)
             self.assertIsNotNone(self.client.get(dbname))
             self.client.__delitem__(dbname, remote=True)
             # Removed from local cache
             self.assertIsNone(self.client.get(dbname))
             # Database removed remotely as well
             try:
-                db = self.client[dbname]
+                _ = self.client[dbname]
```
Comment (alfinkel) [inline tests/unit/db/account_tests.py]: Why change `db` to `_`?

File: tests/unit/db/database_tests.py [hunk 11]
```diff
@@ -315,7 +317,7 @@ def test_get_non_existing_doc_via_getitem(self):
         Test __getitem__ when retrieving a non-existing document
         """
         try:
-            doc = self.db['no_such_doc']
+            _ = self.db['no_such_doc']
```
Comment (alfinkel) [inline tests/unit/db/database_tests.py]: Same question as before.  Why `doc` to `_`?

File: tests/unit/db/unit_t_db_base.py [hunk 12]
```diff
@@ -113,24 +117,22 @@ def setUp(self):
...
-        		self.pwd,
-        		url=self.url,
-        		x_cloudant_user=self.account
-        		)
+            self.account = os.environ.get('CLOUDANT_ACCOUNT')
+            self.user = os.environ.get('DB_USER')
+            self.pwd = os.environ.get('DB_PASSWORD')
+            self.url = os.environ.get(
+                    'DB_URL',
+                    'https://{0}.cloudant.com'.format(self.account))
```
Comment (alfinkel) [inline tests/unit/db/unit_t_db_base.py]: I guess these are also over indented.

File: tests/unit/db/unit_t_db_base.py [hunk 13]
```diff
@@ -113,24 +117,22 @@ def setUp(self):
...
+            self.client = Cloudant(
+                    self.user,
+                    self.pwd,
+                    url=self.url,
+                    x_cloudant_user=self.account)
```
Comment (alfinkel) [inline tests/unit/db/unit_t_db_base.py]: Same, over indented.

File: src/cloudant/database.py (hunk 4 above)
Comment (rredburn) [inline src/cloudant/database.py]: That does not work with pylint 1.6/python 2.7, or at least, I could not get it to work.  pylint would only pass with the override on the same line (a "line" override) vs. above (a "block" override).  The preferred "block" style override works fine with Python 3.5.  I noted this in prior comments -- if you know how to make it work I'm happy to try.  I've moved all PY2 conditionals to _2to3 based on Dave's feedback so this exact instance is gone at least.

File: src/cloudant/database.py (hunk 5 above)
Comment (rredburn) [inline src/cloudant/database.py]: The problem is we either have to relax your pylintrc constraint of 1000 lines per file or, as I've done here, remove newlines.  I did not want to arbitrarily relax your testing standards so opted to flatten a few method calls as noted in my original comments.  Please advise.

File: src/cloudant/document.py (hunk 1 above)
Comment (rredburn) [inline src/cloudant/document.py]: agreed, all PY2 conditional imports have been moved to _2to3 module as suggested.

File: src/cloudant/replicator.py (hunk 8 above)
Comment (rredburn) [inline src/cloudant/replicator.py]: Pycharm and I found the scope a little unnerving as repl_doc is used in the outer parent function (review the larger context).  Changing the inner function variable to arepl_doc removes potential confusion on if you intend to manipulate the outer variable or create a new one.  I will undo the edit if desired.

File: src/cloudant/result.py (hunk 9 above)
Comment (rredburn) [inline src/cloudant/result.py]: in py3, keys() returns an iterator which would then have to be forced into list form for "in" testing.  Rather than do that twice back to back, I chose to do it once and make it a set to improve performance, ie. fewer list constructions and an O(1) vs. O(n) membership test.

File: tests/unit/db/account_tests.py (hunk 10 above)
Comment (rredburn) [inline tests/unit/db/account_tests.py]: ah, remove complaints about unused variables to declutter pycharm warnings.  I can revert or simply remove the assignment altogether, let me know.

File: src/cloudant/database.py (hunk 4 above)
Comment (alfinkel) [inline src/cloudant/database.py]: Fair enough.  pylint is often a pain.

File: src/cloudant/database.py (hunk 5 above)
Comment (alfinkel) [inline src/cloudant/database.py]: Sorry, I had no idea that we exceeded 1000 lines.  I think it would be fine if you upped the `max-module-lines` to 2000 in the [pylintrc](https://github.com/cloudant/python-cloudant/blob/master/pylintrc#L209).  Most of the lines in the modules are docstring anyway.

File: src/cloudant/replicator.py (hunk 8 above)
Comment (alfinkel) [inline src/cloudant/replicator.py]: Good point.  Leave it as you have it.

File: src/cloudant/result.py (hunk 9 above)
Comment (alfinkel) [inline src/cloudant/result.py]: Thanks for the clarification.

File: tests/unit/db/account_tests.py (hunk 10 above)
Comment (alfinkel) [inline tests/unit/db/account_tests.py]: We've been a bit more lax with the test code as far as standards go.  I had left the `db` in there for it to be a bit clearer as to the purpose of the test.  

I guess my main concern initially was that some changes that were being made in this PR did not seem like they were py2 to py3 related. After reading your responses to some of the earlier comments I realize now that that is not the case. If you ran pycharm on the entire testing code base then obviously you made this type of "fix" everywhere pycharm had a similar issue so I am fine with you leaving this as it is throughout.

**Comment by alfinkel** [comment]
@rredburn:  Thanks for the clear responses.  I think we're pretty close.  We have a few topics left to address:

- Some minor formatting comments.
- Adding the Python 3 test to Travis (https://github.com/cloudant/python-cloudant/pull/91#issuecomment-178658173)
- Running the tests against Cloudant. (https://github.com/cloudant/python-cloudant/pull/91#issuecomment-178659909)
- Upping the limit on module size for pylint (https://github.com/cloudant/python-cloudant/pull/91#discussion_r51758450)

**Comment by rredburn** [comment]
@alfinkel, @evansde77  :  I believe I've addressed all comments to date.  The latest commit includes:

* all requested code edits
* Python 3.5 Travis check,  passes both 2.7 and 3.5
* Cloudant tests run and pass both 2.7 and 3.5.  One thing to note, I did get infrequent, seemingly random, unrepeatable fails when testing against a free Cloudant 30 day trial account.  Such fails occurred in both 2.7 and 3.5 and have no relation to Python code issues that I see.  I believe they are due to race conditions occurring with unreliable network or shared server performance variations (replication tests in particular seem to pass or hang/timeout randomly between consecutive runs).  It certainly wouldn't hurt to run the tests in a dedicated and reliable environment to confirm no issues.

Please review and mark additional edits as needed.  Thanks.

**Comment by alfinkel** [comment]
I think the ball is in my court now.  I have a few tests that I want to run before moving forward.

**Comment by alfinkel** [comment]
+1...

I've merged the contents of the commits from this PR branch over top of a rebased branch in this repo and issued PR #97.  Closing this one.
